### ISSUE TITLE
Integrate GPT 4o without TTS/STT

### ISSUE THREAD

**Issue by clemlesne** [microsoft/call-center-ai#210]
OpenAI GPT 4o model supports both in and out of text, image and audio. Understanding is finer than usual STT > model > TTS approach because the model has direct access to user behavior, emotions, etc.

Is there a way to use Communication Services and receive the raw audio flow, bypassing the STT step?
```[tasklist]
### Tasks
- [ ] https://github.com/microsoft/call-center-ai/issues/316
- [ ] https://github.com/microsoft/call-center-ai/issues/317
```

**Comment by Qwatro55** [comment]
I'm also interested in this question.

**Comment by agentverket** [comment]
What about response time?
What about costs?
Can you stream data?

**Comment by clemlesne** [comment]
I know I know :) OpenAI APIs are not yet available:

- https://community.openai.com/t/what-will-the-gpt-4o-audio-api-look-like/754242
- https://community.openai.com/t/gpt-4o-chat-completion-with-audio-response/752149/6

Plus, Communication Services APIs are not yet available to use with raw audio stream.

If you have ideas, don't hesitate!

**Comment by JunJD** [comment]
m

**Comment by clemlesne** [comment]
Audio streaming is now available with Communication Services!

https://learn.microsoft.com/en-us/azure/communication-services/how-tos/call-automation/audio-streaming-quickstart?pivots=programming-language-python

**Comment by ngoanpv** [comment]
Realtime API now support speech to speech from OpenAI.
https://platform.openai.com/docs/guides/realtime/overview 

I would like to explore more and add this feature to this project @clemlesne

**Comment by clemlesne** [comment]
We're working on it!

![Clémence is working](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExMG54OWFzdGJlcTN2c2xlM3F3MXM4amZpaml3dHV0MzY5MjRpNmlzOCZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/JIX9t2j0ZTN9S/giphy.gif)

**Comment by clemlesne** [comment]
Changes merged (5ca7ac2cfc89e83558c0056f328c69b74eaa84f3)!

`gpt-4o-realtime` has not been implemented due to service limitations and model hallucinations. But, a relatively similar UX has been achieved (23503dc2ffdcabd4fdfa663bf06dcbfc0e91dbd0) by manually integrating the VAD algorithm and using Azure AI Speech. More details here: https://github.com/microsoft/call-center-ai/issues/317#issuecomment-2490724994.

PR and feedbacks appreciated :)

**Comment by clemlesne** [comment]
Closing for now, as it seems functionally solved. May re-open in the future to continue the work on `gpt-4o-realtime` once it will be more stable and less prone to strong hallucinations.
